hazelcast/src/main/java/com/hazelcast/replicatedmap/impl/record/KeySetIteratorFactory.java

```diff
@@ -49,8 +48,9 @@ private KeySetIterator(Iterator<Map.Entry<K, ReplicatedRecord<K, V>>> iterator)
         @Override
         public boolean hasNext() {
             while (iterator.hasNext()) {
-                entry = iterator.next();
+                Map.Entry<K, ReplicatedRecord<K, V>> entry = iterator.next();
                 if (testEntry(entry)) {
+                    nextEntry = entry;
```

Comment:
This is the actual fix: set the `nextEntry` field only, when the `entry` is valid (no tombstone).